lightning-invoice/src/lib.rs

```diff
@@ -949,49 +1002,54 @@ impl PositiveTimestamp {
 		if unix_seconds > SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME {
 			Err(CreationError::TimestampOutOfBounds)
 		} else {
-			Ok(PositiveTimestamp(UNIX_EPOCH + Duration::from_secs(unix_seconds)))
+			Ok(PositiveTimestamp(Duration::from_secs(unix_seconds)))
 		}
 	}
 
+	#[cfg(feature = "std")]
 	/// Create a new `PositiveTimestamp` from a `SystemTime` with a corresponding unix timestamp in
-	/// the Range `0...SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME`, otherwise return a
+	/// the range `0...SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME`, otherwise return a
 	/// `CreationError::TimestampOutOfBounds`.
 	pub fn from_system_time(time: SystemTime) -> Result<Self, CreationError> {
-		if time
-			.duration_since(UNIX_EPOCH)
-			.map(|t| t.as_secs() <= SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME)
-			.unwrap_or(true)
-			{
-				Ok(PositiveTimestamp(time))
+		let duration_res = time.duration_since(SystemTime::UNIX_EPOCH);
+		if let Ok(duration) = duration_res {
+			if duration.as_secs() <= SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME {
+				Ok(PositiveTimestamp(duration))
 			} else {
+					Err(CreationError::TimestampOutOfBounds)
+			}
+		} else {
+			Err(CreationError::TimestampOutOfBounds)
+		}
+	}
+
+	/// Create a new `PositiveTimestamp` from a `Duration` since the UNIX epoch in
+	/// the range `0...SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME`, otherwise return a
+	/// `CreationError::TimestampOutOfBounds`.
+	pub fn from_duration_since_epoch(duration: Duration) -> Result<Self, CreationError> {
+		if duration.as_secs() <= SYSTEM_TIME_MAX_UNIX_TIMESTAMP - MAX_EXPIRY_TIME {
+			Ok(PositiveTimestamp(duration))
+		} else {
 			Err(CreationError::TimestampOutOfBounds)
 		}
 	}
 
 	/// Returns the UNIX timestamp representing the stored time
 	pub fn as_unix_timestamp(&self) -> u64 {
```

Comment:
We should be consistent about terminology (`duration_since_epoch` vs `unix_timestamp`) as well as in the builder.

Reply:
This is a public API though.  Should we change the name even so?

Reply:
(note that the PR uses "duration since epoch" when it talks about a `Duration` object, and "unix timestamp" when it talks about a `u64`)

Reply:
(I updated some comments where the terminology was unclear)

Reply:
Thanks, LGTM